src/Common/src/Interop/OSX/Interop.CoreFoundation.CFProxy.cs

```diff
@@ -0,0 +1,145 @@
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Diagnostics;
+using System.Runtime.InteropServices;
+using System.Text;
+using Microsoft.Win32.SafeHandles;
+
+using CFRunLoopSourceRef = System.IntPtr;
+
+internal static partial class Interop
+{
+    internal static partial class CoreFoundation
+    {
+        [DllImport(Libraries.CFNetworkLibrary)]
+        internal static extern SafeCFDictionaryHandle CFNetworkCopySystemProxySettings();
+
+        [DllImport(Libraries.CFNetworkLibrary)]
+        internal static extern SafeCFArrayHandle CFNetworkCopyProxiesForURL(SafeCreateHandle url, SafeCFDictionaryHandle proxySettings);
+
+        internal delegate void CFProxyAutoConfigurationResultCallback(IntPtr client, IntPtr proxyList, IntPtr error);
+
+        [DllImport(Libraries.CFNetworkLibrary)]
+        internal static extern CFRunLoopSourceRef CFNetworkExecuteProxyAutoConfigurationURL(
+            IntPtr proxyAutoConfigURL,
+            SafeCreateHandle targetURL,
+            CFProxyAutoConfigurationResultCallback cb,
+            ref CFStreamClientContext clientContext);
+
+        [DllImport(Libraries.CFNetworkLibrary)]
+        internal static extern CFRunLoopSourceRef CFNetworkExecuteProxyAutoConfigurationScript(
+            IntPtr proxyAutoConfigurationScript,
+            SafeCreateHandle targetURL,
+            CFProxyAutoConfigurationResultCallback cb,
+            ref CFStreamClientContext clientContext);
+
+        [StructLayout(LayoutKind.Sequential)]
+        internal struct CFStreamClientContext
+        {
+            public IntPtr Version;
+            public IntPtr Info;
+            public IntPtr Retain;
+            public IntPtr Release;
+            public IntPtr CopyDescription;
+        }
+
+        internal class CFProxy
+        {
+            private SafeCFDictionaryHandle _dictionary;
+
+            internal static readonly string kCFProxyTypeAutoConfigurationURL;
+            internal static readonly string kCFProxyTypeAutoConfigurationJavaScript;
+            internal static readonly string kCFProxyTypeFTP;
+            internal static readonly string kCFProxyTypeHTTP;
+            internal static readonly string kCFProxyTypeHTTPS;
+            internal static readonly string kCFProxyTypeSOCKS;
+
+            private static readonly IntPtr kCFProxyAutoConfigurationJavaScriptKey;
+            private static readonly IntPtr kCFProxyAutoConfigurationURLKey;
+            private static readonly IntPtr kCFProxyHostNameKey;
+            private static readonly IntPtr kCFProxyPasswordKey;
+            private static readonly IntPtr kCFProxyPortNumberKey;
+            private static readonly IntPtr kCFProxyTypeKey;
+            private static readonly IntPtr kCFProxyUsernameKey;
+
+            static CFProxy()
+            {
+                IntPtr lib = Interop.Libdl.dlopen(Interop.Libraries.CFNetworkLibrary, Interop.Libdl.RTLD_LAZY);
```

Comment:
Could we use the new NativeLibrary APIs instead of direct dlopen usage?

Reply:
Maybe. I started looking into it, but I ended up compiling and testing the code outside of CoreFX where it was not available. There's very little information about `NativeLibrary` overall and I wasn't sure if the API is stable already.